File: mgk-frontend/packages/sdk/src/pda.ts
import { PublicKey } from '@solana/web3.js';

export const SEEDS = {
  PORTFOLIO: 'portfolio',
  INSTRUMENT: 'instrument',
  VAULT: 'vault',
  REGISTRY: 'registry',
  BATCH: 'batch',
  COMMITMENT: 'commitment',
  BOOK: 'book',
} as const;

function seedBytes(s: string): Uint8Array {
  return new Uint8Array(Buffer.from(s, 'utf-8'));
}

function u16Le(value: number): Uint8Array {
  const buf = new Uint8Array(2);
  new DataView(buf.buffer).setUint16(0, value, true);
  return buf;
}

function u64Le(value: bigint): Uint8Array {
  const buf = new Uint8Array(8);
  new DataView(buf.buffer).setBigUint64(0, value, true);
  return buf;
}

export function derivePortfolioPda(
  user: PublicKey,
  programId: PublicKey,
): readonly [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [seedBytes(SEEDS.PORTFOLIO), user.toBytes()],
    programId,
  );
}

export function deriveInstrumentPda(
  instrumentId: number,
  programId: PublicKey,
): readonly [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [seedBytes(SEEDS.INSTRUMENT), u16Le(instrumentId)],
    programId,
  );
}

export function deriveVaultPda(
  programId: PublicKey,
): readonly [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [seedBytes(SEEDS.VAULT)],
    programId,
  );
}

export function deriveRegistryPda(
  programId: PublicKey,
): readonly [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [seedBytes(SEEDS.REGISTRY)],
    programId,
  );
}

export function deriveBatchPda(
  batchId: bigint,
  programId: PublicKey,
): readonly [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [seedBytes(SEEDS.BATCH), u64Le(batchId)],
    programId,
  );
}

export function deriveCommitmentPda(
  batchId: bigint,
  user: PublicKey,
  salt: bigint,
  programId: PublicKey,
): readonly [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [
      seedBytes(SEEDS.COMMITMENT),
      u64Le(batchId),
      user.toBytes(),
      u64Le(salt),
    ],
    programId,
  );
}

export function deriveBookPda(
  instrumentId: number,
  programId: PublicKey,
): readonly [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [seedBytes(SEEDS.BOOK), u16Le(instrumentId)],
    programId,
  );
}

File: programs/perps-core/src/pda.rs
use pinocchio::pubkey::{find_program_address, Pubkey};

pub const PORTFOLIO_SEED: &[u8] = b"portfolio";
pub const INSTRUMENT_SEED: &[u8] = b"instrument";
pub const VAULT_SEED: &[u8] = b"vault";
pub const REGISTRY_SEED: &[u8] = b"registry";
pub const BATCH_SEED: &[u8] = b"batch";
pub const COMMITMENT_SEED: &[u8] = b"commitment";

pub fn derive_portfolio_pda(user: &Pubkey, program_id: &Pubkey) -> (Pubkey, u8) {
    find_program_address(&[PORTFOLIO_SEED, user.as_ref()], program_id)
}

pub fn derive_instrument_pda(instrument_id: u16, program_id: &Pubkey) -> (Pubkey, u8) {
    find_program_address(&[INSTRUMENT_SEED, &instrument_id.to_le_bytes()], program_id)
}

pub fn derive_vault_pda(program_id: &Pubkey) -> (Pubkey, u8) {
    find_program_address(&[VAULT_SEED], program_id)
}

pub fn derive_registry_pda(program_id: &Pubkey) -> (Pubkey, u8) {
    find_program_address(&[REGISTRY_SEED], program_id)
}

pub fn derive_batch_pda(batch_id: u64, program_id: &Pubkey) -> (Pubkey, u8) {
    find_program_address(&[BATCH_SEED, &batch_id.to_le_bytes()], program_id)
}

pub fn derive_commitment_pda(
    batch_id: u64,
    user: &Pubkey,
    nonce: u64,
    program_id: &Pubkey,
) -> (Pubkey, u8) {
    find_program_address(
        &[
            COMMITMENT_SEED,
            &batch_id.to_le_bytes(),
            user.as_ref(),
            &nonce.to_le_bytes(),
        ],
        program_id,
    )
}
